### DISCUSSION TITLE
Async Thread Pool Best Practice

### DISCUSSION THREAD

**Discussion by JKomoroski** [dropwizard/dropwizard#8562] [Q&A]
I have a Dropwizard application which is largely IO bound, so I am planning to make use of `CompletableFuture.runAsync` and `CompletableFuture.supplyAsync` to wrap some of that IO work and get things running concurrently. I've had some great success so far in my prototypes.

The default behavior is to send that work to the common ForkJoinPool workers, but I see the option to pass an Executor if desired. Given that the common pool will vary based on the environment, I am looking for some advice on what a good default practice is in dropwizard.

The three options I see are 
1) Use the default, adjust my environment to compensate
2) Create my own Thread Pool (Requires More Baseline Memory, maybe slower starts)
3) Perhaps I could tap into the DW/Jersey thread pool? Not sure what the implications or what access would be here. 

Does anyone have some production experience and advice to shed?

Java 11 & Java 17, Dropwizard 2.1 & Dropwizard 4 if it matters.

**Comment by Clsan** [comment]
- We use `io.dropwizard.metrics.jetty11.InstrumentedQueuedThreadPool`. And we pass this to CompletableFuture's executor.
- By using it, you can set allocate threads by each job's load.
- Also, you can monitor each job by `com.codahale.metrics.MetricRegistry`.

**Reply by JKomoroski** [reply]
So far, I have settled with creating my own ExecutorService by tapping into the dropwizard environment lifecycle:
```
        return environment.lifecycle()
                .executorService("my-async-pool")
                // managed executors must have a bound to scale out the active thread pool
                // https://github.com/dropwizard/dropwizard/pull/834
                .workQueue(new LinkedBlockingQueue<>(config.getAsyncPoolSize() * 128))
                .maxThreads(config.getAsyncPoolSize())
                .minThreads(config.getAsyncPoolSize())
                .build();
```

I can pass in value from my configuration class at start up. My environment lends itself toward statically sized thread pools. The lifecycle executor service builder seems to create something very similar to what you are suggesting.

**Reply by zUniQueX** [reply]
@JKomoroski I think going with the JDK executors will be better for this use case than using an instance of Jetty's `QueuedThreadPool`. Jetty does many other things in this class, for example adding support for virtual threads via an other executor. If you only need one executor for your asynchronous tasks, that class could imply too much overhead.

Using the dropwizard Jetty `QueuedThreadPool` will interfere with request handling and probably lead to a higher latency accepting a new connection. You probably want to accept connections fast for lower response times in endpoints that don't need asynchronous computation.

If you need metrics for your async executor, you can provide a custom `InstrumentedThreadFactory` to the lifecycle's builder method.
